compiler/pipes/calc-bad-vars.cpp

```diff
@@ -115,13 +115,297 @@ struct FuncCallGraph {
+      }
+    }
+  }
+
+  // on error (colored chain breaks some rule), we want to find an actual chain of calling
+  // this is launched only on error, that's why we don't care about performance and just use bfs
+  // todo think about using such approach for fork chains, throws chains, etc (and drop fork_prev, wait_prev, throws_reason from FunctionData)
+  std::vector<FunctionPtr> find_callstack_between_two_functions_bfs(FunctionPtr from, FunctionPtr target, const std::unordered_set<FunctionPtr> &shouldnt_appear) {
+    std::unordered_map<FunctionPtr, int> visited_level;
+    std::queue<FunctionPtr> bfs_queue;
+
+    bfs_queue.push(from);
+    visited_level[from] = 0;
+
+    while (!bfs_queue.empty()) {
+      FunctionPtr cur = bfs_queue.front();
+      bfs_queue.pop();
+      int next_level = visited_level[cur] + 1;
+
+      if (cur == target) {
+        break;
+      }
+
+      for (FunctionPtr called : call_graph.graph[cur]) {
+        if (shouldnt_appear.find(called) == shouldnt_appear.end() && visited_level.emplace(called, next_level).second) {
+          bfs_queue.push(called);
+        }
+      }
+    }
+
+    if (visited_level.find(target) == visited_level.end()) {    // couldn't find, just return [from, target]
+      return {from, target};
+    }
+
+    std::vector<FunctionPtr> callstack;
+    callstack.emplace_back(target);
+    for (FunctionPtr cur = target; cur != from;) {
+      int prev_level = visited_level[cur] - 1;
+      for (FunctionPtr callee : call_graph.rev_graph[cur]) {
+        auto found = visited_level.find(callee);
+        if (found != visited_level.end() && found->second == prev_level) {
+          callstack.push_back(callee);
+          cur = callee;
+          break;
+        }
+      }
+      kphp_assert(visited_level.find(cur)->second == prev_level);
+    }
+
+    std::reverse(callstack.begin(), callstack.end());
+    return callstack;
+  }
+
+  void show_error_on_rule_broken(const CallstackOfColoredFunctions &color_callstack, const function_palette::PaletteRule &rule) {
+    kphp_assert(rule.is_error());
+    if (shown_errors.size() > max_shown_errors_count) {
+      return;
+    }
+
+    auto vector = color_callstack.as_vector();    // f1, f2, f3 — all of them are colored, and their chain breaks the rule
+    std::vector<FunctionPtr> full_callstack;      // will be: src_main -> ... -> f1 -> ... -> f2 -> ... -> f3
+    for (int i = 0; i < vector.size() - 1; ++i) {
+      FunctionPtr cur = vector[i], next = vector[i+1];
+      std::unordered_set<FunctionPtr> shouldnt_appear;
+      kphp_assert(cur->next_with_colors);
+      for (FunctionPtr exclude : *cur->next_with_colors) {
+        if (exclude != next && exclude != cur) {
+          shouldnt_appear.insert(exclude);
+        }
+      }
+      auto next_callstack_part = find_callstack_between_two_functions_bfs(cur, next, shouldnt_appear);
+      full_callstack.insert(full_callstack.end(), next_callstack_part.begin(), next_callstack_part.end() - 1);
```

Comment:
```suggestion
      full_callstack.insert(full_callstack.end(), next_callstack_part.begin(), next_callstack_part.rbegin());
```

Reply:
Hmm, it seems we can't use `rbegin` here because it's a reverse iterator.

Reply:
Hmm, sorry. Additionally we need to call `base()` to get the corresponding base iterator, so it will be like this:
`next_callstack_part.rbegin().base()`

Reply:
ok, sorry, let's just leave it as is :)